worker/worker.go

```diff
@@ -64,8 +81,15 @@ func (w *Pool) StartWorkers() {
 					metrics.Timing("worker_processing_duration_milliseconds", (now.Sub(batchReadTime).Milliseconds())/lenBatch, "worker="+workerName)
 					metrics.Timing("server_processing_latency_milliseconds", (now.Sub(request.TimeConsumed)).Milliseconds()/lenBatch, fmt.Sprintf("conn_group=%s", request.ConnectionIdentifier.Group))
 				}
-				metrics.Count("kafka_messages_delivered_total", totalErr, fmt.Sprintf("success=false,conn_group=%s", request.ConnectionIdentifier.Group))
-				metrics.Count("kafka_messages_delivered_total", len(request.GetEvents())-totalErr, fmt.Sprintf("success=true,conn_group=%s", request.ConnectionIdentifier.Group))
+
+				for eType, count := range eventCounts {
```

Comment:
Instead of these iterations in the worker, to split the events & errors by type, you can refactor the publisher to produce a `BulkStats` instead of `BulkErrors` where the BulkStats as the eventCounts & eventErrs by type. I'm suggesting this because the producer is already iterating over these events and we should avoid these redundant iterations just to produce the metrics.
The BulkStats can have the map that you have here composing of events & errors by type. In the Producer.Kafka, you can iterate over the errors like this
if errors != null {
    //iterate errors by type
   //populate the BulkStats
}

and finally in the worker you can just use the BulkStats to fetch and report the metrics.

wdyt?

Reply:
Makes sense, we can refactor that way!